web-client/src/views/FileDocument/StateDrivenFileInput.tsx

```diff
@@ -48,6 +57,53 @@ export const StateDrivenFileInput = connect<
 
     const fileOnForm = file || form[fileInputName] || form.existingFileName;
 
+    // Setting the filename here so that we can display it before validation
+    // finishes, otherwise a slow machine might have slight lag and allow the user
+    // to "choose file" again.
+    const [selectedFilename, setSelectedFilename] = useState('');
+
+    const onFileSelectionChange = async (
+      e: React.ChangeEvent<HTMLInputElement>,
+    ) => {
+      setSelectedFilename(e.target?.files?.[0]?.name || '');
```

Comment:
There are a lot of similarities to the implementation in ScanBatchPreview.tsx, and therefore there might be a way to refactor. That said, this PR is already large, so I thought that was best reserved for later.

This isn't a very Cerebral-specific way to do things--i.e., it is not calling a sequence directly but instead passing sequences into functions. Here are my thoughts:

1. The original file validation here was also being done in a "non-Cerebral" way, so this follows the existing precedent
2. There are benefits to pulling into a sequence (consistency) as well as benefits to keeping it distinct (modularity). I can't decide which is a better tradeoff, but the former takes more work. Therefore, I have left things as is.

Reply:
EDIT: Trying to do a lightweight refactor to have the best of both worlds